modules/python/python-persist.c

```diff
@@ -127,3 +161,272 @@ python_format_persist_name(const LogPipe *p, const gchar *module, PythonPersistM
 
   return persist_name;
 }
+
+static PyObject *
+_persist_type_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+  PyPersist *self =(PyPersist *)type->tp_alloc(type, 0);
+
+  GlobalConfig *cfg = main_loop_get_current_config(main_loop_get_instance());
+  self->persist_state = cfg->state;
+
+  return (PyObject *)self;
+}
+
+static int
+_persist_type_init(PyObject *s, PyObject *args, PyObject *kwds)
+{
+  PyPersist *self =(PyPersist *)s;
+  const gchar *persist_name=NULL;
+
+  static char *kwlist[] = {"persist_name", NULL};
+
+  if (! PyArg_ParseTupleAndKeywords(args, kwds, "s", kwlist, &persist_name))
+    return -1;
+
+  if (g_strstr_len(persist_name, -1, SUBKEY_DELIMITER))
+    {
+      // Internally, we will store subkeys as persist_name##subkey
+      PyErr_Format(PyExc_ValueError, "persist name cannot contain " SUBKEY_DELIMITER);
+      return -1;
+    }
+
+  if (!self->persist_name)
+    self->persist_name = g_strdup(persist_name);
+
+  return 0;
+}
+
+static void
+py_persist_dealloc(PyObject *s)
+{
+  PyPersist *self =(PyPersist *)s;
+
+  g_free(self->persist_name);
+  self->persist_name = NULL;
+
+  py_slng_generic_dealloc(s);
+}
+
+static PyMemberDef
+py_persist_type_members[] =
+{
+  {"persist_name", T_STRING, offsetof(PyPersist, persist_name), READONLY},
+  {NULL}
+};
+
+static gchar *
+_build_key(PyPersist *self, const gchar *key)
+{
+  return g_strdup_printf("%s" SUBKEY_DELIMITER "%s", self->persist_name, key);
+}
+
+static gchar *
+_lookup_entry(PyPersist *self, const gchar *key, guint8 *type)
+{
+  gchar *query_key = _build_key(self, key);
+
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(self->persist_state, query_key, &size, &version);
+  if (!handle)
+    {
+      PyErr_Format(PyExc_KeyError, "Persist has no such key: %s", key);
+      g_free(query_key);
+      return NULL;
+    }
+
+  Entry *entry = persist_state_map_entry(self->persist_state, handle);
+  *type = entry->type;
+  gchar *copy = g_strdup(entry->data);
+  persist_state_unmap_entry(self->persist_state, handle);
+
+  g_free(query_key);
+
+  return copy;
+};
+
+static PersistEntryHandle
+_allocate_persist_entry(PersistState *persist_state, const gchar *key, const gchar *value, gsize value_len)
+{
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(persist_state, key, &size, &version);
+  if (handle && size >= value_len)
+    return handle;
+
+  return persist_state_alloc_entry(persist_state, key, value_len);
+}
+
+static gchar *
+_serialize(guint8 type, PyObject *v)
+{
+  if (type == ENTRY_TYPE_STRING)
+    {
+      return g_strdup(_py_get_string_as_string(v));
+    }
+  else if (type == ENTRY_TYPE_LONG)
+    {
+      PyObject *as_str = PyObject_Str(v);
```

Comment:
The str() function may fail as it calls the __str__ method.

Reply:
Technically, this should not fail. At this point we are sure `v` is a number. I added a g_assert() just in case.

Reply:
you can derive a new class from the builtin int class and override its __str__ method. exception from _any_ python function is possible.

the addition of the assertion is good enough for me. btw: but you can also get the value out of the int/long object and format it yourself (e.g. PyLong_AsLong())